src/adapter/src/catalog/state.rs

```diff
@@ -0,0 +1,1711 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::net::Ipv4Addr;
+use std::time::Instant;
+
+use anyhow::bail;
+use itertools::Itertools;
+use serde::Serialize;
+use tracing::info;
+
+use mz_audit_log::{EventDetails, EventType, ObjectType, VersionedEvent, VersionedStorageUsage};
+use mz_build_info::DUMMY_BUILD_INFO;
+use mz_catalog::builtin::{Builtin, BuiltinCluster, BuiltinLog, BuiltinSource, BuiltinTable};
+use mz_controller::clusters::{
+    ClusterStatus, ManagedReplicaLocation, ProcessId, ReplicaConfig, ReplicaLocation,
+};
+use mz_controller_types::{ClusterId, ReplicaId};
+use mz_expr::MirScalarExpr;
+use mz_ore::cast::CastFrom;
+use mz_ore::collections::CollectionExt;
+use mz_ore::now::{to_datetime, EpochMillis, NOW_ZERO};
+use mz_ore::soft_assert;
+use mz_repr::adt::mz_acl_item::PrivilegeMap;
+use mz_repr::namespaces::{
+    INFORMATION_SCHEMA, MZ_CATALOG_SCHEMA, MZ_INTERNAL_SCHEMA, MZ_TEMP_SCHEMA, PG_CATALOG_SCHEMA,
+};
+use mz_repr::role_id::RoleId;
+use mz_repr::{GlobalId, RelationDesc};
+use mz_sql::catalog::{
+    CatalogCluster, CatalogClusterReplica, CatalogConfig, CatalogDatabase,
+    CatalogError as SqlCatalogError, CatalogItem as SqlCatalogItem, CatalogItemType, CatalogRole,
+    CatalogSchema, EnvironmentId, IdReference, SessionCatalog, SystemObjectType, TypeReference,
+};
+use mz_sql::names::{
+    CommentObjectId, DatabaseId, FullItemName, FullSchemaName, ItemQualifiers, ObjectId,
+    PartialItemName, QualifiedItemName, QualifiedSchemaName, RawDatabaseSpecifier,
+    ResolvedDatabaseSpecifier, ResolvedIds, SchemaId, SchemaSpecifier, SystemObjectId,
+};
+use mz_sql::plan::{CreateViewPlan, Params, Plan};
+use mz_sql::session::user::MZ_SYSTEM_ROLE_ID;
+use mz_sql::session::vars::{SystemVars, Var, VarInput};
+use mz_sql_parser::ast::QualifiedReplica;
+use mz_storage_types::connections::inline::{
+    ConnectionResolver, InlinedConnection, IntoInlineConnection,
+};
+use mz_transform::Optimizer;
+
+use crate::catalog::{
+    AwsPrincipalContext, BuiltinTableUpdate, Catalog, CatalogEntry, CatalogItem, Cluster,
+    ClusterConfig, ClusterReplica, ClusterReplicaProcessStatus, ClusterReplicaSizeMap, CommentsMap,
+    Database, DefaultPrivileges, Error, ErrorKind, Index, Role, Schema, View,
+    LINKED_CLUSTER_REPLICA_NAME, SYSTEM_CONN_ID,
+};
+use crate::client::ConnectionId;
+use crate::coord::ConnMeta;
+use crate::session::Session;
+use crate::util::{index_sql, ResultExt};
+use crate::AdapterError;
+
+#[derive(Debug, Clone, Serialize)]
+pub struct CatalogState {
```

Comment:
Can you add a module comment or a doc comment to this struct explaining what `CatalogState` is and how it relates to the what is stored in the Stash? Also maybe call out that we only implement `Serialize` on this struct for the purposes of dumping state, and that the on-disk format in the Stash is protobuf?